src/Storages/ExternalStream/Kafka/WriteBufferFromKafka.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <atomic>
+#include <librdkafka/rdkafka.h>
+
+#include "IO/BufferWithOwnMemory.h"
+#include "IO/WriteBuffer.h"
+
+namespace DB
+{
+
+using RdKafkaTopicPtr = std::unique_ptr<rd_kafka_topic_t, void (*)(rd_kafka_topic_t *)>;
+
+class WriteBufferFromKafka : public BufferWithOwnMemory<WriteBuffer>
+{
+private:
+    struct State
+    {
+        std::atomic_size_t outstandings;
+        std::atomic_size_t acked;
+        std::atomic_int errorCode;
+    };
+
+public:
+    WriteBufferFromKafka(const Poco::Logger * logger, RdKafkaTopicPtr topic_, size_t buffer_size = DBMS_DEFAULT_BUFFER_SIZE);
+
+    ~WriteBufferFromKafka() override = default;
+
+    void onDrMsg(const rd_kafka_message_t *);
+    bool fullyAcked() const;
+    rd_kafka_resp_err_t deliveryError() const;
```

Comment:
We shall translate `rd_kafka_resp_err_t` to ErrorCodes to avoid the very low level error propagate the modular boundry and we have some util in `KafkaLog` already.

Reply:
I can't use `ErrorCodes` here because I need to use `rd_kafka_err2str` to get the error message from the error code.